src/redis-benchmark.c

```diff
@@ -1771,6 +1823,162 @@ int main(int argc, char **argv) {
+ * *returns* of values are non-zero nanosecond cost in comparison to just *not*
+ * doing those things. So we won't when we need to benchmark crc64 of short
+ * strings, or when we need to benchmark the <100 nanosecond timing of crc
+ * combining. */
+
+        init_nstime();
+
+        struct timespec init_start, init_end;
+        NSTIME(&init_start);
+        /* need to benchmark the combine init first, before we get to crc anything */
+        crc64_combine(
+            UINT64_C(0xdeadbeefdeadbeef),
+            UINT64_C(0xfeebdaedfeebdaed),
+            INIT_SIZE,
+            RPOLY, 64);
+        NSTIME(&init_end);
+
+        crc64_init();
+
+        // We want to hash about 1 gig of data in total, looped, to get a good
+        // idea of our performance.
+        uint64_t expect = 0;
+
+        if (config.crc64_test_size > 0) {
+            genBenchmarkRandomData((char*)data, config.crc64_test_size);
+            // warm up the cache
+            set_crc64_cutoffs(config.crc64_test_size+1, config.crc64_test_size+1);
+            expect = crc64(0, data, config.crc64_test_size);
+            if (config.csv) printf("algorithm,buffer,performance,crc64_matches\n");
+        } else {
+            // just need some nonzero bytes here
+            genBenchmarkRandomData((char*)(&expect), sizeof(uint64_t));
+        }
+
+        // We have nanoseconds, and we need megs/second from bytes. Great,
+        // just a multiply and a division.
+        // By sampling over a range of --crc <sizes>, you can determine the
+        // best values for for setting CRC64_[DUAL|TRI]_CUTOFF on your machine.
+        // NOTE: we needed nstime here for handling timing of processing --crc
+        // values < 4k or so, as they can complete in under 1 microsecond on
+        // some platforms.
+
+        // save some copy / paste
+#define bench(WHICH) \
+        uint64_t min_##WHICH = INIT_SIZE; \
+        uint64_t hash_##WHICH; \
+        for (long long i=passes; i > 0; i--) { \
+            struct timespec original_start, original_end; \
+            NSTIME(&original_start); \
+            hash_##WHICH = crc64(0, data, csize); \
+            NSTIME(&original_end); \
+            uint64_t delta = nstimediff(&original_start, &original_end); \
+            min_##WHICH = min_##WHICH < delta ? min_##WHICH : delta; \
+        } if (config.csv) \
+        printf("%s,%" PRIu64 ",%" PRIu64 ",%d\n", #WHICH, csize, (1000 * csize) / min_##WHICH, \
+            hash_##WHICH == hash_crc_1byte); \
+        else printf("test size=%" PRIu64 " algorithm=%s %" PRIu64 " M/sec matches=%d\n", \
+            csize, #WHICH, (1000 * csize) / min_##WHICH, \
+            hash_##WHICH == hash_crc_1byte)
+
+        if (config.crc64_test_size > 0) {
+            for (uint64_t csize = (uint64_t)config.crc64_test_size;
+                 csize > 3;
+                 csize -= (csize >> 2) + (csize >> 3) + (csize >> 6)) {
```

Comment:
I'm sure there is valid explanations for all of this stuff, but I have no idea why weren't doing half of this math/comparisons.

Reply:
This is mostly to get a nice range of testing points in one go. Will simplify.